File: ios_app/iphone/DaytimePickerSliderView.swift
import SwiftUI
import shared

private let hPadding = H_PADDING + 1

// Slider line a little wider to nice ui
private let sliderInternalPadding = 3.0
private let circleSize = 24.0
private let circleDefaultOffset = hPadding - (circleSize / 2) + sliderInternalPadding

// No size of tick, but for absolute positioned view
private let tickNoteWidthDp = 20.0

struct DaytimePickerSliderView: View {

    let daytimeUi: DaytimeUi
    let onChange: (DaytimeUi) -> Void

    var body: some View {

        let sliderUi = DaytimePickerSliderUi(
            hour: daytimeUi.hour,
            minute: daytimeUi.minute
        )

        VStack {

            SliderView(
                tickIdx: sliderUi.hour.toInt(),
                ticks: sliderUi.hourTicks,
                stepTicks: sliderUi.hourStepSlide.toInt(),
                onChange: { hour in
                    let newDaytimeUi = DaytimeUi(
                        hour: hour.toInt32(),
                        minute: daytimeUi.minute
                    )
                    onChange(newDaytimeUi)
                }
            )
            .padding(.top, 4)

            SliderView(
                tickIdx: sliderUi.minute.toInt(),
                ticks: sliderUi.minuteTicks,
                stepTicks: sliderUi.minuteStepSlide.toInt(),
                onChange: { minute in
                    let newDaytimeUi = DaytimeUi(
                        hour: daytimeUi.hour,
                        minute: minute.toInt32()
                    )
                    onChange(newDaytimeUi)
                }
            )
            .padding(.top, 14)
            .padding(.bottom, 8)
        }
    }
}

private struct SliderView: View {

    let tickIdx: Int
    let ticks: [DaytimePickerSliderUi.Tick]
    let stepTicks: Int
    let onChange: (Int) -> Void

    // 0 also used as "no value"
    @State private var sliderXPx = 0.0
    // Not exact but approximate `slide width` / `ticks size`
    @State private var tickAxmPx = 0.0

    @State private var circleOffsetAnimation = 0.0

    var body: some View {

        ZStack(alignment: .topLeading) {

            ZStack {

                ZStack {
                }
                .frame(minWidth: 0, maxWidth: .infinity)
                .frame(height: 4)
                .background(GeometryReader { geometry -> Color in
                    myAsyncAfter(0.01) {
                        let frame = geometry.frame(in: CoordinateSpace.global)
                        let newSliderXPx = frame.origin.x
                        // "ticks.size - 1" to make first and last sticks on edges
                        let newTickPx = geometry.size.width / (ticks.count - 1).toDouble()
                        if !(sliderXPx == newSliderXPx && tickAxmPx == newTickPx) {
                            sliderXPx = newSliderXPx
                            tickAxmPx = newTickPx
                        }
                    }
                    return Color.clear
                })
                .padding(.horizontal, sliderInternalPadding)
                .background(roundedShape.fill(c.dividerBg))
                .padding(.horizontal, hPadding)

                //
                // Circle
                // Ignore init animation

                if sliderXPx > 0 {

                    let circleOffset = circleDefaultOffset + (tickAxmPx * tickIdx.toDouble())

                    HStack {

                        HStack {

                            Text(ticks[tickIdx].text)
                                .foregroundColor(c.text)
                                .font(.system(size: 12, weight: .medium))
                        }
                        .frame(width: circleSize, height: circleSize, alignment: .center)
                        .background(roundedShape.fill(.blue))
                        .offset(x: circleOffsetAnimation)

                        Spacer()
                    }
                    .animateVmValue(
                        value: circleOffset,
                        state: $circleOffsetAnimation,
                        animation: .spring(response: 0.15)
                    )
                }

                ////
            }
            .frame(height: circleSize)
            .zIndex(1.0)

            ZStack(alignment: .top) {

                ForEachIndexed(ticks) { tickIdx, tick in

                    let tickNoteHalfPx = tickNoteWidthDp / 2
                    let offsetX = (tickAxmPx * tickIdx.toDouble()) - tickNoteHalfPx

                    ZStack(alignment: .top) {

                        if tick.withSliderStick {
                            ZStack {
                            }
                            .frame(width: 1, height: 5)
                            .background(c.dividerFg)
                        }

                        if let sliderStickText = tick.sliderStickText {
                            Text(sliderStickText)
                                .padding(.top, 8)
                                .foregroundColor(.secondary)
                                .font(.system(size: 13, weight: .light))
                        }
                    }
                    .offset(x: offsetX)
                    .frame(width: tickNoteWidthDp)
                }
            }
            .zIndex(0.0)
            .padding(.top, 18)
            .padding(.horizontal, hPadding + sliderInternalPadding)
        }
        .highPriorityGesture(gesture)
    }

    var gesture: some Gesture {

        DragGesture(minimumDistance: 0, coordinateSpace: .global)
            .onChanged { value in
                let slideXPosition = value.location.x - sliderXPx
                let stepPx = tickAxmPx * stepTicks.toDouble()

                let newIdx = DaytimePickerSliderUi.companion.calcSliderTickIdx(
                    ticksSize: ticks.count.toInt32(),
                    stepTicks: stepTicks.toInt32(),
                    slideXPosition: Float(slideXPosition),
                    stepPx: Float(stepPx)
                )

                if tickIdx != newIdx {
                    onChange(newIdx.toInt())
                }
            }
    }
}

File: android_app/src/main/java/me/timeto/app/ui/daytime_picker/DaytimePickerSliderView.kt
package me.timeto.app.ui.daytime_picker

import android.view.MotionEvent
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.VisibilityThreshold
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.pointer.motionEventSpy
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.layout.positionInWindow
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import me.timeto.app.misc.Haptic
import me.timeto.app.ui.H_PADDING
import me.timeto.app.ui.Padding
import me.timeto.app.ui.VStack
import me.timeto.app.ui.ZStack
import me.timeto.app.ui.c
import me.timeto.app.ui.onePx
import me.timeto.app.ui.pxToDp
import me.timeto.app.ui.roundedShape
import me.timeto.shared.DaytimeUi
import me.timeto.shared.DaytimePickerSliderUi

private val hPadding = H_PADDING + 1.dp

// Slider line a little wider to nice ui
private val sliderInternalPadding = 3.dp
private val circleSize = 22.dp
private val circleDefaultOffset = hPadding - (circleSize / 2) + sliderInternalPadding
private val circleAnimation = spring(
    visibilityThreshold = Dp.VisibilityThreshold,
    stiffness = Spring.StiffnessHigh,
)

// No size of tick, but for absolute positioned view
private val tickNoteWidthDp: Dp = 16.dp

private val allowedMotionEventActions = setOf(
    MotionEvent.ACTION_DOWN,
    MotionEvent.ACTION_MOVE,
    MotionEvent.ACTION_UP,
)

@Composable
fun DaytimePickerSliderView(
    daytimeUi: DaytimeUi,
    modifier: Modifier,
    onChange: (DaytimeUi) -> Unit,
) {

    val sliderUi = DaytimePickerSliderUi(
        hour = daytimeUi.hour,
        minute = daytimeUi.minute,
    )

    VStack(
        modifier = modifier,
    ) {

        SliderView(
            tickIdx = sliderUi.hour,
            ticks = sliderUi.hourTicks,
            stepTicks = sliderUi.hourStepSlide,
            onChange = { newHour ->
                onChange(daytimeUi.copy(hour = newHour))
            },
        )

        Padding(vertical = 8.dp)

        SliderView(
            tickIdx = sliderUi.minute,
            ticks = sliderUi.minuteTicks,
            stepTicks = sliderUi.minuteStepSlide,
            onChange = { newMinute ->
                onChange(daytimeUi.copy(minute = newMinute))
            },
        )
    }
}

@OptIn(ExperimentalComposeUiApi::class)
@Composable
private fun SliderView(
    tickIdx: Int,
    ticks: List<DaytimePickerSliderUi.Tick>,
    stepTicks: Int,
    onChange: (Int) -> Unit,
) {

    // 0 also used as "no value"
    val sliderXPx = remember { mutableIntStateOf(0) }
    // Not exact but approximate `slide width` / `ticks size`
    val tickAxmPx = remember { mutableFloatStateOf(0f) }

    ZStack(
        modifier = Modifier
            .motionEventSpy { event ->
                if (event.action !in allowedMotionEventActions)
                    return@motionEventSpy

                val slideXPosition = event.x - sliderXPx.intValue
                val stepPx = tickAxmPx.floatValue * stepTicks

                val newIdx = DaytimePickerSliderUi.calcSliderTickIdx(
                    ticksSize = ticks.size,
                    stepTicks = stepTicks,
                    slideXPosition = slideXPosition,
                    stepPx = stepPx,
                )

                if (tickIdx != newIdx) {
                    onChange(newIdx)
                    Haptic.shot()
                }
            }
    ) {

        ZStack(
            modifier = Modifier
                .zIndex(1f)
                .height(circleSize),
        ) {

            ZStack(
                modifier = Modifier
                    .align(Alignment.Center)
                    .padding(horizontal = hPadding)
                    .fillMaxWidth()
                    .height(4.dp)
                    .clip(roundedShape)
                    .background(c.gray5)
                    .padding(horizontal = sliderInternalPadding)
                    .onGloballyPositioned { coords ->
                        val newSliderXPx = coords.positionInWindow().x.toInt()
                        // "ticks.size - 1" to make first and last sticks on edges
                        val newTickPx = coords.size.width.toFloat() / (ticks.size - 1)
                        if (sliderXPx.intValue == newSliderXPx && tickAxmPx.floatValue == newTickPx)
                            return@onGloballyPositioned
                        sliderXPx.intValue = newSliderXPx
                        tickAxmPx.floatValue = newTickPx
                    },
            )

            //
            // Circle
            // Ignore init animation

            if (sliderXPx.intValue > 0) {

                val circleOffset = remember(tickIdx) {
                    derivedStateOf {
                        circleDefaultOffset + pxToDp((tickAxmPx.floatValue * tickIdx).toInt()).dp
                    }
                }
                val circleOffsetAnimation = animateDpAsState(
                    targetValue = circleOffset.value,
                    animationSpec = circleAnimation,
                )

                ZStack(
                    modifier = Modifier
                        .offset(x = circleOffsetAnimation.value)
                        .size(circleSize)
                        .clip(roundedShape)
                        .background(c.blue),
                    contentAlignment = Alignment.Center,
                ) {

                    Text(
                        text = ticks[tickIdx].text,
                        modifier = Modifier
                            .offset(y = -onePx * 2),
                        color = c.text,
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Medium,
                    )
                }
            }

            ///
        }

        ZStack(
            modifier = Modifier
                .zIndex(0f)
                .padding(top = 16.dp)
                .padding(horizontal = hPadding + sliderInternalPadding)
                .fillMaxWidth(),
        ) {

            ticks.forEachIndexed { tickIdx, tick ->

                ZStack(
                    modifier = Modifier
                        .offset {
                            val tickNoteHalfPx = (tickNoteWidthDp.value * this.density) / 2
                            IntOffset(((tickAxmPx.floatValue * tickIdx) - tickNoteHalfPx).toInt(), 0)
                        }
                        .width(tickNoteWidthDp),
                    contentAlignment = Alignment.TopCenter,
                ) {

                    if (tick.withSliderStick) {
                        ZStack(
                            modifier = Modifier
                                .width(1.dp)
                                .height(5.dp)
                                .background(c.gray4),
                        )
                    }

                    val sliderStickText = tick.sliderStickText
                    if (sliderStickText != null) {
                        Text(
                            text = sliderStickText,
                            modifier = Modifier
                                .padding(top = 4.dp),
                            color = c.secondaryText,
                            fontSize = 12.sp,
                            fontWeight = FontWeight.Light,
                        )
                    }
                }
            }
        }
    }
}
